experimental/async_impl.hpp

```diff
@@ -0,0 +1,565 @@
+/*
+ *  Copyright (c) 2020 Intel Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#ifndef async_reduce_impl_hpp
+#define async_reduce_impl_hpp
+
+//#include "oneapi/dpl/pstl/hetero/numeric_impl_hetero.h"
+
+namespace oneapi
+{
+
+namespace dpl
+{
+
+namespace __par_backend_hetero
+{
+
+//template <typename _ExecPolicy>
+using __future_with_tmps = oneapi::dpl::__internal::__future<void> /*<_ExecPolicy>*/;
+
+using __future_base = oneapi::dpl::__internal::__future<void>;
+
+//------------------------------------------------------------------------
+// parallel_stable_sort - async pattern 2.0
+//-----------------------------------------------------------------------
+
+template <typename _ExecutionPolicy, typename _Iterator, typename _Merge, typename _Compare>
+oneapi::dpl::__internal::__enable_if_async_execution_policy<_ExecutionPolicy, __future_with_tmps>
+__parallel_sort_impl_async(_ExecutionPolicy&& __exec, _Iterator __first, _Iterator __last, _Merge __merge,
```

Comment:
Since this is a backend pattern, it could be placed in `parallel_backend_sycl.h`. It would be great if it was merged with existing patterns to not duplicate the code.